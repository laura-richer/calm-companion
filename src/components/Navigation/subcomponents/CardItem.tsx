import React from 'react';
import { Pressable, Text, Image } from 'react-native';
import DropShadow from 'react-native-drop-shadow';
import { CardItemProps } from '../types';
import styles from '../styles';
import { colors } from '~styles/variables';
import { dropShadowStyles, typographyStyles } from '~styles/global';

const CardItem = ({ item, handleChangeScreen }: CardItemProps) => {
  const formattedTitle = item.title.split(' ');

  return (
    <React.Fragment>
      <DropShadow style={[dropShadowStyles.default, styles.cardNavigationItem]}>
        <Pressable
          accessibilityLabel={item.title}
          style={({ pressed }) => [
            styles.cardNavigationItemInner,
            {
              backgroundColor: pressed ? colors.GREY : colors.WHITE,
            },
          ]}
          key={item.id}
          onPress={() => handleChangeScreen(item.id)}
        >
          <Text style={[typographyStyles.textMd, styles.cardNavigationItemText]}>
            {formattedTitle[0]}
            {'\n'}
            {formattedTitle[1]}
          </Text>
          <Image
            alt={`${item.title} icon`}
            accessibilityHint="Icon"
            style={styles.cardNavigationItemImage}
            resizeMode="contain"
            source={item.cardIcon}
          />
        </Pressable>
      </DropShadow>
    </React.Fragment>
  );
};

export default CardItem;
