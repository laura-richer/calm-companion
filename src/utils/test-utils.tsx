import React, { PropsWithChildren } from 'react';
import { NavigationContainer } from '@react-navigation/native';
import { render, RenderOptions } from '@testing-library/react-native';
import { Provider as StoreProvider } from 'react-redux';
import { EnhancedStore, configureStore } from '@reduxjs/toolkit';
import { SafeAreaProvider } from 'react-native-safe-area-context';
import { RootState } from '~store/index';
import navigationReducer, {
  initialState as navigationInitialState,
} from '~store/slices/navigation';
import quickHelpReducer, { initialState as quickHelpInitialState } from '~store/slices/quickHelp';

type DeepPartial<T> = {
  [K in keyof T]?: T[K] extends object ? DeepPartial<T[K]> : T[K];
};

interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
  preloadedState?: Partial<RootState>;
  store?: EnhancedStore;
}

export const renderWithProviders = (
  ui: React.ReactElement,
  { preloadedState = {}, store = createTestStore(preloadedState) }: ExtendedRenderOptions = {}
) => {
  const wrapper = ({ children }: PropsWithChildren<{}>): JSX.Element => {
    return (
      <StoreProvider store={store}>
        <SafeAreaProvider>
          <NavigationContainer>{children}</NavigationContainer>
        </SafeAreaProvider>
      </StoreProvider>
    );
  };

  return {
    ...render(ui, {
      wrapper,
    }),
  };
};

const _createStore = (preloadedState: Partial<RootState> = {}): EnhancedStore => {
  return configureStore({
    reducer: {
      navigation: navigationReducer,
      quickHelp: quickHelpReducer,
    },
    preloadedState,
    middleware: getDefaultMiddleware =>
      getDefaultMiddleware({
        immutableCheck: { warnAfter: 256 },
        serializableCheck: { warnAfter: 256 },
      }),
  });
};

export const createTestStore = (preloadedState: DeepPartial<RootState> = {}): EnhancedStore => {
  const state: RootState = {
    // @ts-ignore
    navigation: {
      ...navigationInitialState,
      ...(preloadedState.navigation || {}),
    },
    // @ts-ignore
    quickHelp: {
      ...quickHelpInitialState,
      ...(preloadedState.quickHelp || {}),
    },
  };
  const store: EnhancedStore = _createStore(state);
  const origDispatch = store.dispatch;
  store.dispatch = jest.fn(origDispatch);
  return store;
};
