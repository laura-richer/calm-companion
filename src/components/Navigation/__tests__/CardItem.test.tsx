import React from 'react';
import { screen } from '@testing-library/react-native';
import { renderWithProviders } from '~utils/test-utils';
import CardItem from '../subcomponents/CardItem';

const mockCardItem = {
  id: 'QuickHelp',
  component: jest.fn(),
  title: 'Quick Help',
  tabIcon: require('~assets/images/tabIcons/quick-help.png'),
  cardIcon: require('~assets/images/screenHeaders/quick-help.png'),
};

const _renderDefault = () => {
  return renderWithProviders(<CardItem item={mockCardItem} handleChangeScreen={() => {}} />);
};

describe('Card item', () => {
  beforeEach(() => {
    _renderDefault();
  });

  test('matches snapshot', () => {
    expect(screen.toJSON()).toMatchSnapshot();
  });

  test('renders card title', () => {
    const title = screen.getByLabelText('Quick Help');
    expect(title).toBeOnTheScreen();
  });

  test('renders card icon', () => {
    const title = screen.getByAccessibilityHint('Icon');
    expect(title).toBeOnTheScreen();
  });
});
