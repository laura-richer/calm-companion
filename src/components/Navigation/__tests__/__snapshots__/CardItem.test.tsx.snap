// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card item matches snapshot 1`] = `
<View
  style={
    [
      {
        "shadowColor": "#8b8490",
        "shadowOffset": {
          "height": 2,
          "width": 0,
        },
        "shadowOpacity": 0.25,
        "shadowRadius": 8,
      },
      {
        "flexBasis": "47%",
        "marginBottom": 20,
      },
    ]
  }
>
  <View
    accessibilityLabel="Quick Help"
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      [
        {
          "borderRadius": 8,
          "padding": 10,
        },
        {
          "backgroundColor": "#ffffff",
        },
      ]
    }
  >
    <Text
      style={
        [
          {
            "fontFamily": "Poppins-Bold",
            "fontSize": 14,
          },
          {
            "color": "#3f7b82",
            "maxWidth": 100,
          },
        ]
      }
    >
      Quick
      

      Help
    </Text>
    <Image
      accessibilityHint="Icon"
      alt="Quick Help icon"
      resizeMode="contain"
      source={
        {
          "testUri": "../../../src/assets/images/screenHeaders/quick-help.png",
        }
      }
      style={
        {
          "marginLeft": "auto",
          "maxWidth": 75,
          "opacity": 0.25,
        }
      }
    />
  </View>
</View>
`;
